src/tests/settings.test.ts

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { ThemeType } from '../style/themes'
+import {
+  defaultSettings,
+  getNetworkName,
+  getStoredSettings,
+  loadStoredSettings,
+  networkEndpoints,
+  saveStoredSettings,
+  updateStoredSettings
+} from '../utils/settings'
+
+const mockSettings = {
+  general: {
+    theme: 'light' as ThemeType,
+    walletLockTimeInMinutes: 999,
+    discreetMode: false,
+    passwordRequirement: false
+  },
+  network: {
+    nodeHost: 'https://node',
+    explorerApiHost: 'https://explorer-backend',
+    explorerUrl: 'https://explorer'
+  }
+}
+
+it('Should return the network name if all settings match exactly', () => {
+  expect(getNetworkName(networkEndpoints.localhost)).toEqual('localhost'),
+    expect(getNetworkName(networkEndpoints.testnet)).toEqual('testnet'),
+    expect(getNetworkName(networkEndpoints.mainnet)).toEqual('mainnet'),
+    expect(getNetworkName({ nodeHost: '', explorerApiHost: '', explorerUrl: '' })).toEqual('custom'),
+    expect(
+      getNetworkName({
+        ...networkEndpoints.mainnet,
+        nodeHost: 'https://mainnet-wallet.alephium2.org'
+      })
+    ).toEqual('custom')
+})
+
+it('Should load settings from local storage', () => {
+  expect(loadStoredSettings()).toEqual(defaultSettings)
+
+  localStorage.setItem('theme', 'pink')
+  expect(localStorage.getItem('theme')).toEqual('pink')
+  expect(loadStoredSettings().general.theme).toEqual('pink')
```

Comment:
Looking at this, I'm wondering if we should maybe split the `loadStoredSettings` into two explicit functions:

- `migrateDeprecatedSettings` => which reuse most of the code found in our current `loadStoredSettings`
- `loadStoredSettings` => a rename of `returnSettingsObject`
- `loadAndMigrateStoredSettings` => `loadStoredSettings()` + `migrateDeprecatedSettings()`

The tests would be clearer like that, as we could test the migration and the loading process individually. WDYT?

Reply:
At the moment we call the `loadStoredSettings` in 4 distinct places in our app:
- `src/components/Address.tsx` (this should be replaced by getting the settings from the global context
- `src/contexts/global.tsx`, once to set the `settings` property of the `initialGlobalContext` object and once to set the default value of the state.
- `src/utils/api-clients.ts` (IMO, the `settings` argument of the `createClient` function should not be optional. That way we don't need to call the `loadStoredSettings` there and we only call it in the global context file.

My opinion is that we should have only 2 functions:
1. `migrateDeprecatedSettings`, which migrates any deprecated settings found in local storage and returns the final settings object. This function could be used only in one place: to set the `settings` property of the `initialGlobalContext` object. This way, it would be called exactly 1 time when the app runs.
2. `loadStoredSettings`, which just loads the stored settings (same what `returnSettingsObject` currently does). This function would be used only 1 one place: to set the default value of the state.

Moreover, I would delete `getStoredSettings` since it's not used anywhere.

Reply:
I full agree with the general idea of simplifying settings loading and avoid unnecessary calls. I'm not so happy with [how](https://github.com/alephium/alephium-wallet/commit/41a131322db5e1a49e07464df4959e4ba28b1a5c#r67647699) the migration method is called though. I'd make it more explicit, and migrate the settings before passing them in the context.

A proposition: we could have a `migrationRequired` flag in the setting object, set by a "test method" called asap on app launch. I would make sure to call this method as early as possible, either in the `contexts/global.tsx`file or even in the good old `index.tsx` , so that we are 100% sure that any following access to the settings is safe (ie. settings are migrated, and in a correct state).
Once the migration is done, the `migrationRequired` is set to `false`, and no more calls are done, until the "migration test method" is changed by a subsequent app update.

WDYT?

Reply:
yeah, I like this approach, I'll work on it tomorrow! Thanks for taking the time!

Reply:
@mvaivre I have just pushed a commit that takes your feedback into consideration for a better solution.

https://github.com/alephium/alephium-wallet/pull/131/commits/8d75b2140ec5bc656b371b85f9f04524f0ca0b4c